core/src/main/java/com/spotify/missinglink/datamodel/MethodDescriptors.java

```diff
@@ -29,17 +33,59 @@ private MethodDescriptors() {
   }
 
   public static MethodDescriptor fromDesc(String desc, String name) {
-    Type type = Type.getMethodType(desc);
+    final MethodKey key = new MethodKey(name, desc);
+    return methodDescriptorCache.computeIfAbsent(key, MethodDescriptors::newDescriptor);
+  }
+
+  private static MethodDescriptor newDescriptor(MethodKey key) {
+    Type type = Type.getMethodType(key.desc);
 
-    List<TypeDescriptor> params = ImmutableList.copyOf(type.getArgumentTypes()).stream()
+    List<TypeDescriptor> params = Arrays.stream(type.getArgumentTypes())
         .map(Type::getDescriptor)
         .map(TypeDescriptors::fromRaw)
         .collect(toList());
 
     return new MethodDescriptorBuilder()
         .returnType(TypeDescriptors.fromRaw(type.getReturnType().getDescriptor()))
-        .name(name)
+        .name(key.name)
         .parameterTypes(ImmutableList.copyOf(params))
         .build();
   }
+
+  private static final Map<MethodKey, MethodDescriptor> methodDescriptorCache = new HashMap<>();
```

Comment:
Petter is correct - the goal was to avoid instantiating anything at all if there is already an existing MethodDescriptor that matches the inputs.